src/main/java/me/zeroeightsix/kami/module/modules/player/PacketLogger.kt

```diff
@@ -1,100 +1,305 @@
-import kotlinx.coroutines.launch
 import me.zeroeightsix.kami.KamiMod
+import me.zeroeightsix.kami.event.events.ConnectionEvent
 import me.zeroeightsix.kami.event.events.PacketEvent
-import me.zeroeightsix.kami.mixin.extension.pitch
-import me.zeroeightsix.kami.mixin.extension.yaw
+import me.zeroeightsix.kami.mixin.extension.*
 import me.zeroeightsix.kami.module.Category
 import me.zeroeightsix.kami.module.Module
-import me.zeroeightsix.kami.util.text.MessageSendHelper
-import me.zeroeightsix.kami.util.threads.defaultScope
+import me.zeroeightsix.kami.util.text.MessageSendHelper.sendChatMessage
+import me.zeroeightsix.kami.util.threads.onMainThread
 import me.zeroeightsix.kami.util.threads.safeListener
-import net.minecraft.network.play.client.CPacketPlayer
-import net.minecraft.network.play.client.CPacketPlayerDigging
+import net.minecraft.network.play.client.*
+import net.minecraft.network.play.server.*
 import net.minecraftforge.fml.common.gameevent.TickEvent
-import java.io.File
+import java.io.*
 import java.text.SimpleDateFormat
 import java.util.*
+import kotlin.ConcurrentModificationException
 
 internal object PacketLogger : Module(
     name = "PacketLogger",
     description = "Logs sent packets to a file",
     category = Category.PLAYER
 ) {
-    private val append by setting("Append", false)
-    private val clear = setting("Clear", false)
+    private val showClientTicks by setting("ShowClientTicks", true)
+    private val logInChat by setting("LogInChat", false)
+    private val packetsFrom by setting("LogPacketsFrom", Type.BOTH)
+    private val ignoreKeepAlive by setting("IgnoreKeepAlive", true)
+    private val ignoreChunkLoading by setting("IgnoreChunkLoading", true)
+    private val ignoreUnknown by setting("IgnoreUnknownPackets", false)
+    private val ignoreChat by setting("IgnoreChat", true)
 
-    private val file = File("${KamiMod.DIRECTORY}packet_logger.txt")
-    private val lines = Collections.synchronizedList(ArrayList<String>())
-    private val sdf = SimpleDateFormat("HH:mm:ss.SSS")
+    private val sdf = SimpleDateFormat("HH-mm-ss_SSS")
+    private val sdflog = SimpleDateFormat("HH:mm:ss.SSS")
+    private var filename = ""
+    private val lines = ArrayList<String>()
+
+    private var start = 0L
+    private var last = 0L
+
+    enum class Type {
+        CLIENT, SERVER, BOTH
+    }
 
     init {
         onEnable {
-            if (append) read()
+            if (mc.player == null) {
+                disable()
+                return@onEnable
+            }
+            onMainThread {
+                start = System.currentTimeMillis()
+            }
+
+            filename = "KAMIBluePackets_${sdf.format(Date())}.csv"
+            lines.add("From,Packet Name,Time Since Start (ms),Time Since Last (ms),Data\n")
+            sendChatMessage("$chatName started.")
         }
 
         onDisable {
-            write()
+            if (mc.player != null) write()
+            sendChatMessage("$chatName stopped.")
```

Comment:
The stop message is entirely unnecessary when toggle messages already exist for each module